src/module_1/module_1_meteo_api.py

```diff
@@ -1,7 +1,240 @@
+openmeteo = openmeteo_requests.Client(session = retry_session)
+
+API_URL = "https://climate-api.open-meteo.com/v1/climate"
+
+COORDINATES = {
+    "Madrid": {"latitude": 40.416775, "longitude": -3.703790},
+    "London": {"latitude": 51.507351, "longitude": -0.127758},
+    "Rio": {"latitude": -22.906847, "longitude": -43.172896},
+}
+
+VARIABLES = "temperature_2m_mean, precipitation_sum, soil_moisture_0_to_10cm_mean"
+
+
+""" Given the response to the API call, prints the according error. """
+def handle_api_error(response):
+    if response.status_code == 401:
+        print("Invalid API key. Check credentials.")
+    
+    elif response.status_code == 404:
+        print("Location not found.")
+    
+    else:
+        print("API error, status code: ", response.status_code)
+        print("API error, reason: ", response.reason)
+
+
+
+""" Given a set of parameters, calls the Meteo API. If the call is successful, it returns the corresponding json;
+otherwise, it returns None, handling the corresponding error."""
+def call_API(params):
+    # call the Meteo API
+
+    try:
+        response = requests.get(API_URL, params)
+    
+        if response.status_code == 200:
+            print("API request has been successful")
+            return response.json()
+        
+        elif response.status_code == 429:
+            print("Rate limit exceeded. Please try again later.")
+            return None
+        else:
+            handle_api_error(response) 
+            return None
+        
+    except:
+        print(f"API Request failed: {requests.RequestException}")
+        return None
+    
+
+""" Given a city, defines the parameters and calls the API. 
+If the call is successful, it returns the corresponding daily dataframe; otherwise, it tries reading the
+corresponding csv file."""
+def get_data_meteo_api(city):
+
+    latitude = COORDINATES[city]['latitude']
+    longitude = COORDINATES[city]['longitude']
+
+    params = {
+	"latitude": latitude,
+	"longitude": longitude,
+	"start_date": "1950-01-01",   # from 1950 to 2050
+	"end_date": "2050-12-31",
+	"models": "EC_Earth3P_HR", # One of the models with all weather variables available
+	"daily": ["temperature_2m_mean", "precipitation_sum", "soil_moisture_0_to_10cm_mean"] # Data we want to extract
+    }
+
+    # Call the API
+    responses = call_API(params)
+
+    file_path = os.path.join(OUTPUT_DIR, f'{city}_daily_data.csv')
+
+    if responses:     
+        daily_json = get_daily_data(responses, city)
+        # Transform to dataframe
+        daily_df = pd.DataFrame(daily_json)
+    
+        # Save it in a csv only if there is not already one
+        if not os.path.exists(file_path):
+            daily_df.to_csv(file_path, index=False)
+        
+        return daily_df
+    
+    else:
+        try:
+            print("Failed to get API data. Loading data from CSV.")
+            return pd.read_csv(file_path)
```

Comment:
Why this dependency? I see it as a source of bugs. What if somebody modify the file?